File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    List<Integer>[] edges;
    int n;
    public int solution(int n, int[][] edge) {
        this.n = n;
        edges = new ArrayList[n + 1];
        for (int i = 0; i <= n; i++)
            edges[i] = new ArrayList<>();
        for (int[] e : edge) {
            edges[e[0]].add(e[1]);
            edges[e[1]].add(e[0]);
        }
        return bfs(1);
    }
    
    int bfs(int start) {
        Queue<Integer> q = new LinkedList<>();
        q.offer(start);
        int[] dist = new int[n + 1];
        dist[start] = 1;
        
        while (!q.isEmpty()) {
            Integer curr = q.poll();
            
            for (Integer next : edges[curr]) {
                if (dist[next] == 0) {
                    dist[next] = dist[curr] + 1;
                    q.offer(next);
                }
            }
        }
        
        int max = 0;
        for (int d : dist)
            if (d > max)
                max = d;
        int count = 0;
        for (int d : dist)
            if (d == max)
                count++;
        return count;
    }
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
#include <iostream>

using namespace std;

vector<int> dist(20001, 50001);

void bfs(int n, int start, vector<vector<int>>& edge) {
    queue<int> q;
    dist[start] = 0;
    q.push(start);
    
    while (!q.empty()) {
        int now = q.front(); q.pop();
        
        for (auto& next : edge[now]) {
            if (dist[next] == 50001) {
                dist[next] = dist[now] + 1;
                q.push(next);
            }
        }
    }
}

int solution(int n, vector<vector<int>> edge) {
    vector<vector<int>> graph(20001);
    for (auto& e : edge) {
        graph[e[0]].push_back(e[1]);
        graph[e[1]].push_back(e[0]);
    }
    bfs(n, 1, graph);
    int answer = 0, count = 0;
    int maxLength = *max_element(dist.begin() + 1, dist.begin() + n + 1);
    for (int d : dist)
        if (d == maxLength)
            answer++;
    return answer;
}
